### PR TITLE
Versioning tools for independent and preview releases

### PR DESCRIPTION
**Pull request by cormacpayne**
## Description

Resolves issue https://github.com/Azure/azure-powershell/issues/5147

- Add `VersionController` project
    - Determines and applies correct version bump to a changed module
    - Validates version bump to a module
- Add `Tools.Common` project
    - Contains shared classes and content between `StaticAnalysis` analyzers and `VersionController`
- Refactor `StaticAnalysis`
    - Use the classes from `Tools.Common`
    - Update the `BreakingChangeAnalyzer` with fixes to type and parameter set comparison
    - Update how assemblies are loaded
- Refactor DataFactories and DataFactoryV2 projects
    - Split the projects into two separate folders for consistency and tooling compliance
- Add script to update version of  `AzureRM`
- Add script to verify correct assemblies are loaded to `build.proj`

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md).**
- [ ] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/preview/documentation/help-generation.md).**
- [ ] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/preview/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#testing-guidelines)
- [ ] Pull request includes test coverage for the included changes.
- [ ] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [ ] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://github.com/Azure/azure-powershell/wiki/PowerShell-Cmdlet-Design-Guidelines#supportsshouldprocess).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [ ] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [ ] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by cormacpayne** [comment]
@azuresdkci test this please

**Comment by MiYanni** [comment]
@markcowl Nudge 👨‍⚕️

**Comment by MiYanni** [comment]
@cormacpayne Can you squash up them committles to a nice, small, commitlet?

**Comment by cormacpayne** [comment]
@MiYanni 🐧

File: tools/StaticAnalysis/BreakingChangeAnalyzer/TypeMetadataHelper.cs [hunk 1]
```diff
@@ -133,6 +134,111 @@ public TypeMetadataHelper()
...
+            TypeMetadata newType,
+            ReportLogger<BreakingChangeIssue> issueLogger)
+        {
+            var oldMethods = oldType.Methods;
+            var newMethods = newType.Methods;
+
+            foreach (var oldMethod in oldMethods)
+            {
+                bool found = false;
+                var candidateMethods = newMethods.Where(m => string.Equals(m.Name, oldMethod.Name)).ToList();
```
Comment (markcowl) [inline tools/StaticAnalysis/BreakingChangeAnalyzer/TypeMetadataHelper.cs]: Are we considering method return type as well?

File: tools/StaticAnalysis/BreakingChangeAnalyzer/TypeMetadataHelper.cs [hunk 2]
```diff
@@ -133,6 +134,111 @@ public TypeMetadataHelper()
...
+            }
+        }
+
+        public void CompareConstructorMetadata(
+            CmdletMetadata cmdlet,
+            TypeMetadata oldType,
+            TypeMetadata newType,
+            ReportLogger<BreakingChangeIssue> issueLogger)
+        {
+            var oldConstructors = oldType.Constructors;
```
Comment (markcowl) [inline tools/StaticAnalysis/BreakingChangeAnalyzer/TypeMetadataHelper.cs]: Other than return types, seems like we should be able to reuse the same code here.

File: tools/StaticAnalysis/HelpAnalyzer/HelpAnalyzer.cs [hunk 3]
```diff
@@ -83,17 +88,18 @@ public void Analyze(IEnumerable<string> scopes)
                     var dirs = Directory.EnumerateDirectories(directory);
                     if (dirs == null || !dirs.Any((d) => string.Equals(Path.GetFileName(d), "help", StringComparison.OrdinalIgnoreCase)))
                     {
```
Comment (markcowl) [inline tools/StaticAnalysis/HelpAnalyzer/HelpAnalyzer.cs]: let's just remove block altogether

File: tools/Tools.Common/Models/ModuleMetadata.cs [hunk 4]
```diff
@@ -52,14 +52,47 @@ private void InitializeTypeDictionary()
...
+
+        public override bool Equals(Object obj)
+        {
+            var other = obj as ModuleMetadata;
+            if (other == null)
+            {
+                return false;
+            }
+
+            var flag = true;
```
Comment (markcowl) [inline tools/Tools.Common/Models/ModuleMetadata.cs]: ```<nit>```this should have a more descriptive name, like modulesEqual```</nit>```

File: tools/Tools.Common/Models/ParameterMetadata.cs [hunk 5]
```diff
@@ -62,5 +62,41 @@ public class ParameterMetadata
...
+        public override bool Equals(Object obj)
+        {
+            var other = obj as ParameterMetadata;
+            if (other == null)
+            {
+                return false;
+            }
+
+            var flag = true;
+
```
Comment (markcowl) [inline tools/Tools.Common/Models/ParameterMetadata.cs]: clean up the vertical spaces in this file - no reason for veritcal spaces except after ending brackets

File: tools/Tools.Common/Models/ParameterMetadata.cs [hunk 6]
```diff
@@ -62,5 +62,41 @@ public class ParameterMetadata
...
+            thisNames.Add(this.Name, true);
+            this.AliasList.ForEach(a => thisNames.Add(a, true));
+
+            var otherNames = new Dictionary<string, bool>();
+            otherNames.Add(other.Name, true);
+            other.AliasList.ForEach(a => otherNames.Add(a, true));
+
+            flag &= thisNames.Count == otherNames.Count;
+
+            flag &= this.ValidateRangeMin == other.ValidateRangeMin;
```
Comment (markcowl) [inline tools/Tools.Common/Models/ParameterMetadata.cs]: You can make this a single expression, starting each line with a new && clause

File: tools/Tools.Common/Models/ParameterMetadata.cs [hunk 7]
```diff
@@ -62,5 +62,41 @@ public class ParameterMetadata
...
+        /// <returns>True if the two objects are equal, false otherwise.</returns>
+        public override bool Equals(Object obj)
+        {
+            var other = obj as ParameterMetadata;
+            if (other == null)
+            {
+                return false;
+            }
+
+            var flag = true;
```
Comment (markcowl) [inline tools/Tools.Common/Models/ParameterMetadata.cs]: Use cmdletsEaul or similar descriptive name

File: tools/Tools.Common/Models/ParameterSetMetadata.cs [hunk 8]
```diff
@@ -37,6 +34,42 @@ public class ParameterSetMetadata
...
+        public override bool Equals(Object obj)
+        {
+            var other = obj as ParameterSetMetadata;
+            if (other == null)
+            {
+                return false;
+            }
+
+            var flag = true;
+
```
Comment (markcowl) [inline tools/Tools.Common/Models/ParameterSetMetadata.cs]: remove extra vertical spaces

File: tools/Tools.Common/Models/ParameterSetMetadata.cs [hunk 9]
```diff
@@ -37,6 +34,42 @@ public class ParameterSetMetadata
...
+                {
+                    return false;
+                }
+
+                flag &= thisParameter.Equals(otherParameter);
+            }
+
+            flag &= this.Parameters.Count == other.Parameters.Count;
+
+            return flag;
```
Comment (markcowl) [inline tools/Tools.Common/Models/ParameterSetMetadata.cs]: more descriptive variable name

File: tools/Tools.Common/Models/ParameterSetMetadata.cs [hunk 10]
```diff
@@ -66,9 +99,34 @@ public class Parameter
...
+        /// <returns>True if the two objects are equal, false otherwise.</returns>
+        public override bool Equals(Object obj)
+        {
+            var other = obj as Parameter;
+            if (other == null)
+            {
+                return false;
+            }
+
+            var flag = true;
```
Comment (markcowl) [inline tools/Tools.Common/Models/ParameterSetMetadata.cs]: same comments as above

File: tools/Tools.Common/Models/TypeMetadata.cs [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+        /// <returns>True if the two objects are equal, false otherwise.</returns>
+        public override bool Equals(Object obj)
+        {
+            var other = obj as TypeMetadata;
+            if (other == null)
+            {
+                return false;
+            }
+
+            var flag = true;
```
Comment (markcowl) [inline tools/Tools.Common/Models/TypeMetadata.cs]: vertical space / more descriptive identifier

File: tools/Tools.Common/Models/TypeMetadata.cs [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+                    return false;
+                }
+
+                foreach (var thisMethod in this.Methods)
+                {
+                    bool found = false;
+                    var candidateMethods = other.Methods.Where(m => string.Equals(m.Name, thisMethod.Name)).ToList();
+                    foreach (var otherMethod in candidateMethods)
+                    {
+                        var thisParameters = thisMethod.Parameters;
```
Comment (markcowl) [inline tools/Tools.Common/Models/TypeMetadata.cs]: Are we checking return types as well?

File: tools/Tools.Common/Models/TypeMetadata.cs [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+            {
+                if (this.Constructors.Count != other.Constructors.Count)
+                {
+                    return false;
+                }
+
+                foreach (var thisConstructor in this.Constructors)
+                {
+                    bool found = false;
+                    foreach (var otherConstructor in other.Constructors)
```
Comment (markcowl) [inline tools/Tools.Common/Models/TypeMetadata.cs]: can we not duplicate this code - this seems the same as the method code.

File: tools/VersionController/Models/VersionBumper.cs [hunk 14]
```diff
@@ -0,0 +1,348 @@
...
+            _metadataHelper = new VersionMetadataHelper(_fileHelper);
+        }
+
+        /// <summary>
+        /// Bump the version in all necessary files.
+        /// </summary>
+        public void BumpAllVersions()
+        {
+            var moduleName = _fileHelper.ModuleName;
+
```
Comment (markcowl) [inline tools/VersionController/Models/VersionBumper.cs]: remove extra vertical space

File: tools/VersionController/Models/VersionBumper.cs [hunk 15]
```diff
@@ -0,0 +1,348 @@
...
+        /// <returns>The updated version after the bump has been applied.</returns>
+        private string GetBumpedVersion()
+        {
+            var moduleName = _fileHelper.ModuleName;
+            var splitVersion = _oldVersion.Split('.').Select(v => int.Parse(v)).ToArray();
+            var versionBump = _metadataHelper.GetVersionBumpUsingSerialized();
+            if (string.Equals(moduleName, "AzureRM.Profile"))
+            {
+                var commonCodeVersionBump = _metadataHelper.GetVersionBumpForCommonCode();
+                if (commonCodeVersionBump == Version.MAJOR)
```
Comment (markcowl) [inline tools/VersionController/Models/VersionBumper.cs]: I think we should throw if this is the case

File: tools/VersionController/Models/VersionBumper.cs [hunk 16]
```diff
@@ -0,0 +1,348 @@
...
+                    versionBump = Version.MAJOR;
+                }
+                else if (commonCodeVersionBump == Version.MINOR && versionBump == Version.PATCH)
+                {
+                    versionBump = Version.MINOR;
+                }
+            }
+
+            // PATCH update for preview modules (0.x.x or x.x.x-preview)
+            if (splitVersion[0] == 0 || _isPreview)
```
Comment (markcowl) [inline tools/VersionController/Models/VersionBumper.cs]: let's just set VersionBump to Revision in tis case and use the code below to set tthe actual version

File: tools/VersionController/Models/VersionMetadataHelper.cs [hunk 17]
```diff
@@ -0,0 +1,487 @@
...
+            {
+                powershell.AddScript("Save-Module -Name AzureRM.Profile -Repository PSGallery -Path " + outputModuleDirectory);
+                var cmdletResult = powershell.Invoke();
+            }
+
+            try
+            {
+                var commonAssemblies = new List<string>
+                {
+                    Path.Combine(outputModuleDirectory, "Microsoft.Azure.Commands.Common.Authentication.Abstractions.dll"),
```
Comment (markcowl) [inline tools/VersionController/Models/VersionMetadataHelper.cs]: Can we derive these somehow from the psd1 requiredAssemblies?  i.e. Microsoft.*Commands*

File: tools/StaticAnalysis/BreakingChangeAnalyzer/TypeMetadataHelper.cs (hunk 1 above)
Comment (cormacpayne) [reply tools/StaticAnalysis/BreakingChangeAnalyzer/TypeMetadataHelper.cs]: Added `ReturnType` property to `MethodMetadata` and am now checking for breaking changes in the return types when a match (name + parameters) is found.

File: tools/StaticAnalysis/BreakingChangeAnalyzer/TypeMetadataHelper.cs (hunk 2 above)
Comment (cormacpayne) [reply tools/StaticAnalysis/BreakingChangeAnalyzer/TypeMetadataHelper.cs]: Added a `MethodSignature` class that `MethodMetadata` and `ConstructorMetadata` inherit from and refactored these two methods into one to compare method signatures.

File: tools/Tools.Common/Models/TypeMetadata.cs (hunk 12 above)
Comment (cormacpayne) [reply tools/Tools.Common/Models/TypeMetadata.cs]: Added in check for return type.

**Comment by markcowl** [comment]
on demand run here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/687/

**Comment by maddieclayton** [comment]
on demand here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/698/

**Pull request merged**